qiskit_dynamics/solvers/solver_classes.py

```diff
@@ -547,87 +551,63 @@ def format_final_states(y, model, y_input, y0_cls):
-    """Helper function for setting up lists of simulations.
-
-    Transform input signals, t_span, and y0 into lists of the same length.
-    Arguments are given as either lists of valid specifications, or as a singleton of a valid
-    specification. Singletons are transformed into a list of length one, then all arguments
-    are expanded to be the same length as the longest argument max_len:
-        - If len(arg) == 1, it will be repeated max_len times
-        - if len(arg) == max_len, nothing is done
-        - If len(arg) not in (1, max_len), an error is raised
-
-    Args:
-        t_span: Time interval specification.
-        y0: Initial state specification.
-        signals: Signal specification.
-
-    Returns:
-        Tuple: tuple of lists of arguments of the same length, along with a bool specifying whether
-        the arguments specified multiple simulations or not.
-
-    Raises:
-        QiskitError: If the length of any arguments are incompatible, or if any singleton
-        is an invalid shape.
+    """Setup t_span, y0, and signals for doing a list of simulations. Utilizes
+    solver_utils.setup_arg_lists.
     """
 
-    multiple_sims = False
-
-    if signals is None:
-        signals = [signals]
-    elif isinstance(signals, tuple):
-        # single Lindblad
-        signals = [signals]
-    elif isinstance(signals, list) and isinstance(signals[0], tuple):
-        # multiple lindblad
-        multiple_sims = True
-    elif isinstance(signals, list) and isinstance(signals[0], (list, SignalList)):
-        # multiple Hamiltonian signals lists
-        multiple_sims = True
-    elif isinstance(signals, SignalList) or (
-        isinstance(signals, list) and not isinstance(signals[0], (list, SignalList))
-    ):
-        # single Hamiltonian signals list
-        signals = [signals]
-    else:
-        raise QiskitError("Signals specified in invalid format.")
+    def signals_to_list(signals):
+        was_list = False
+        if signals is None:
+            signals = [signals]
+        elif isinstance(signals, tuple):
+            # single Lindblad
+            signals = [signals]
+        elif isinstance(signals, list) and isinstance(signals[0], tuple):
+            # multiple lindblad
+            was_list = True
+        elif isinstance(signals, list) and isinstance(signals[0], (list, SignalList)):
+            # multiple Hamiltonian signals lists
+            was_list = True
+        elif isinstance(signals, SignalList) or (
+            isinstance(signals, list) and not isinstance(signals[0], (list, SignalList))
+        ):
+            # single Hamiltonian signals list
+            signals = [signals]
+        else:
+            raise QiskitError("Signals specified in invalid format.")
 
-    if not isinstance(y0, list):
-        y0 = [y0]
-    else:
-        multiple_sims = True
+        return signals, was_list
 
-    t_span_ndim = nested_ndim(t_span)
+    def y0_to_list(y0):
```

Comment:
the three `to_list` functions can be define outside of `setup_simulation_lists`

Reply:
yeah I think that's a good idea, I'll do that

Reply:
Actually with this change I'm able to get rid of the `setup_simulation_lists`, as it now contains only the call to `setup_args_lists`. I've changed the call to `setup_simulation_lists` to a call directly to `setup_args_lists`, which seems cleaner.